File: LeetCode_Java/BitOperation/338.counting-bits.java
package LeetCode_Java.BitOperation;
/*
 * @lc app=leetcode id=338 lang=java
 *
 * [338] Counting Bits
 */

// @lc code=start
class Solution {
    public int[] countBits(int n) {
        int[] result = new int[n+1];
        for (int i = 0; i < result.length; i++) {
            result[i] = 0;
        }

        for (int i = 0; i < result.length; i++) {
            result[i] = result[i>>1] + (i&1);
        }
        return result; 
    }
    public int[] brutalForce(int n) {
        int[] result = new int[n+1];
        for (int i=0; i < n+1; i++) {
            int count = 0;
            int currentNum = i;
            while (currentNum > 0) {
                count += currentNum&1;
                currentNum >>= 1;
            }
            result[i] = count;
        }
        return result;
    }
}
// @lc code=end

File: LeetCode_Go/BitManipulation/338.counting-bits.go
/*
 * @lc app=leetcode id=338 lang=golang
 *
 * [338] Counting Bits
 */

// @lc code=start

// Given an integer n, return an array ans of length n + 1 such that for each i (0 <= i <= n), ans[i] is the number of 1's in the binary representation of i. (0 <= n <= 10^5)

// use right shift to iterate over the number's bit representation until it becomes 0

func countBits(n int) []int {
	ans := make([]int, n+1)
	for i := 0; i <= n; i++ {
		count := 0
		num := i
		for num != 0 {
			count += num & 1
			num >>= 1
		}
		ans[i] = count
	}
	return ans
}

// @lc code=end

